File: py-basis/src/calculator/indicators.py
"""
Statistical indicators calculation (MA, EMA, Z-Score).
"""

import numpy as np
from typing import List, Dict, Any, Optional
from dataclasses import dataclass
from loguru import logger

# Try to import polars, but have fallback
try:
    import polars as pl
    POLARS_AVAILABLE = True
except ImportError:
    POLARS_AVAILABLE = False
    logger.warning("Polars not available, using NumPy for calculations")


@dataclass
class IndicatorResult:
    """Result of indicator calculations."""
    symbol: str
    timestamp: float
    ma_price: float  # Moving average of price
    ma_basis: float  # Moving average of basis
    ema_basis: float  # Exponential moving average of basis
    z_score: float   # Z-Score of basis


class IndicatorCalculator:
    """Calculator for statistical indicators."""

    def __init__(self, window_size: int = 30):
        """
        Initialize indicator calculator.

        Args:
            window_size: Window size for MA/EMA calculations
        """
        self.window_size = window_size

        # Store basis history per symbol (using native lists for performance)
        self.basis_history: Dict[str, List[float]] = {}
        self.price_history: Dict[str, List[float]] = {}

        # EMA smoothing factor
        self.alpha = 2 / (window_size + 1)

    def add_data(self, symbol: str, basis: float, price: float):
        """
        Add basis and price data for a symbol.

        Args:
            symbol: Trading symbol
            basis: Basis value
            price: Price value
        """
        symbol_key = symbol.upper()

        # Initialize lists if needed
        if symbol_key not in self.basis_history:
            self.basis_history[symbol_key] = []
            self.price_history[symbol_key] = []

        # Add data
        self.basis_history[symbol_key].append(basis)
        self.price_history[symbol_key].append(price)

        # Keep only window_size most recent values
        if len(self.basis_history[symbol_key]) > self.window_size:
            self.basis_history[symbol_key] = self.basis_history[symbol_key][-self.window_size:]
            self.price_history[symbol_key] = self.price_history[symbol_key][-self.window_size:]

    def calculate_indicators(self, symbol: str) -> Optional[IndicatorResult]:
        """
        Calculate indicators for a symbol.

        Args:
            symbol: Trading symbol

        Returns:
            IndicatorResult if enough data, None otherwise
        """
        symbol_key = symbol.upper()

        # Check if we have enough data
        if (symbol_key not in self.basis_history or
            len(self.basis_history[symbol_key]) < self.window_size):
            return None

        try:
            basis_data = self.basis_history[symbol_key]
            price_data = self.price_history[symbol_key]

            # Use Polars if available for optimized calculations
            if POLARS_AVAILABLE and len(basis_data) >= self.window_size:
                return self._calculate_with_polars(symbol_key, basis_data, price_data)
            else:
                return self._calculate_with_numpy(symbol_key, basis_data, price_data)

        except Exception as e:
            logger.error(f"Error calculating indicators for {symbol}: {e}")
            return None

    def _calculate_with_polars(self, symbol: str, basis_data: List[float], price_data: List[float]) -> IndicatorResult:
        """Calculate indicators using Polars (optimized)."""
        # Convert to Polars Series only when needed (performance optimization)
        basis_series = pl.Series("basis", basis_data)
        price_series = pl.Series("price", price_data)

        # Calculate MA (simple moving average)
        ma_basis = basis_series.tail(self.window_size).mean()
        ma_price = price_series.tail(self.window_size).mean()

        # Calculate EMA (exponential moving average)
        # Polars doesn't have built-in EMA for Series, so we calculate manually
        ema_basis = self._calculate_ema_numpy(basis_data)

        # Calculate Z-Score: (current - MA) / StdDev
        current_basis = basis_data[-1] if basis_data else 0
        std_basis = basis_series.tail(self.window_size).std()

        if std_basis > 0:
            z_score = (current_basis - ma_basis) / std_basis
        else:
            z_score = 0.0

        return IndicatorResult(
            symbol=symbol,
            timestamp=np.float64(time.time()),
            ma_price=float(ma_price),
            ma_basis=float(ma_basis),
            ema_basis=float(ema_basis),
            z_score=float(z_score),
        )

    def _calculate_with_numpy(self, symbol: str, basis_data: List[float], price_data: List[float]) -> IndicatorResult:
        """Calculate indicators using NumPy (fallback)."""
        import numpy as np
        import time

        # Convert to numpy arrays for efficient calculation
        basis_array = np.array(basis_data[-self.window_size:])
        price_array = np.array(price_data[-self.window_size:])

        # Calculate MA (simple moving average)
        ma_basis = np.mean(basis_array)
        ma_price = np.mean(price_array)

        # Calculate EMA (exponential moving average)
        ema_basis = self._calculate_ema_numpy(basis_data)

        # Calculate Z-Score: (current - MA) / StdDev
        current_basis = basis_data[-1] if basis_data else 0
        std_basis = np.std(basis_array) if len(basis_array) > 1 else 0

        if std_basis > 0:
            z_score = (current_basis - ma_basis) / std_basis
        else:
            z_score = 0.0

        return IndicatorResult(
            symbol=symbol,
            timestamp=time.time(),
            ma_price=float(ma_price),
            ma_basis=float(ma_basis),
            ema_basis=float(ema_basis),
            z_score=float(z_score),
        )

    def _calculate_ema_numpy(self, data: List[float]) -> float:
        """
        Calculate EMA using NumPy.

        Args:
            data: List of values

        Returns:
            EMA value
        """
        import numpy as np

        if not data:
            return 0.0

        # Use the last window_size values
        window_data = data[-self.window_size:] if len(data) >= self.window_size else data

        # Calculate EMA manually
        weights = np.exp(np.linspace(-1, 0, len(window_data)))
        weights /= weights.sum()

        return np.dot(window_data, weights)

    def get_indicator_history(self, symbol: str, count: int = 100) -> List[IndicatorResult]:
        """
        Get historical indicator data.

        Note: This is inefficient for real-time use - only for occasional queries.
        For real-time, store indicator results as they're calculated.

        Args:
            symbol: Trading symbol
            count: Number of historical points

        Returns:
            List of historical IndicatorResult
        """
        # In a real implementation, you would store calculated indicator results
        # For now, we'll just return empty list
        return []

    def calculate_all_indicators(self) -> Dict[str, IndicatorResult]:
        """
        Calculate indicators for all symbols with sufficient data.

        Returns:
            Dictionary mapping symbol to IndicatorResult
        """
        results = {}
        for symbol in list(self.basis_history.keys()):
            result = self.calculate_indicators(symbol)
            if result:
                results[symbol] = result
        return results


# Import time here to avoid circular import
import time

File: rust-basis/src/calculator/indicators.rs
//! Statistical indicators calculation (MA, EMA, Z-Score).

use std::collections::HashMap;
use polars::prelude::*;
use tracing::{error, warn};
use anyhow::{Result, Context};

/// Result of indicator calculations.
#[derive(Debug, Clone)]
pub struct IndicatorResult {
    /// Trading symbol
    pub symbol: String,
    /// Timestamp
    pub timestamp: u64,
    /// Moving average of price
    pub ma_price: f64,
    /// Moving average of basis
    pub ma_basis: f64,
    /// Exponential moving average of basis
    pub ema_basis: f64,
    /// Z-Score of basis
    pub z_score: f64,
}

/// Calculator for statistical indicators.
pub struct IndicatorCalculator {
    /// Window size for MA/EMA calculations
    window_size: usize,
    /// Basis history by symbol (using native Vec for performance)
    basis_history: HashMap<String, Vec<f64>>,
    /// Price history by symbol (using native Vec for performance)
    price_history: HashMap<String, Vec<f64>>,
    /// EMA smoothing factor
    alpha: f64,
}

impl IndicatorCalculator {
    /// Create a new indicator calculator.
    pub fn new(window_size: usize) -> Self {
        Self {
            window_size,
            basis_history: HashMap::new(),
            price_history: HashMap::new(),
            alpha: 2.0 / (window_size as f64 + 1.0),
        }
    }

    /// Add basis and price data for a symbol.
    pub fn add_data(&mut self, symbol: &str, basis: f64, price: f64) {
        let symbol_key = symbol.to_uppercase();

        // Initialize vectors if needed
        self.basis_history
            .entry(symbol_key.clone())
            .or_insert_with(Vec::new)
            .push(basis);

        self.price_history
            .entry(symbol_key.clone())
            .or_insert_with(Vec::new)
            .push(price);

        // Keep only window_size most recent values
        if let Some(history) = self.basis_history.get_mut(&symbol_key) {
            if history.len() > self.window_size {
                *history = history[history.len() - self.window_size..].to_vec();
            }
        }

        if let Some(history) = self.price_history.get_mut(&symbol_key) {
            if history.len() > self.window_size {
                *history = history[history.len() - self.window_size..].to_vec();
            }
        }
    }

    /// Calculate indicators for a symbol.
    pub fn calculate_indicators(&self, symbol: &str) -> Result<Option<IndicatorResult>> {
        let symbol_key = symbol.to_uppercase();

        // Check if we have enough data
        let basis_data = match self.basis_history.get(&symbol_key) {
            Some(data) if data.len() >= self.window_size => data,
            _ => return Ok(None),
        };

        let price_data = match self.price_history.get(&symbol_key) {
            Some(data) if data.len() >= self.window_size => data,
            _ => return Ok(None),
        };

        // Use Polars for optimized calculations (convert only when needed)
        match self.calculate_with_polars(&symbol_key, basis_data, price_data) {
            Ok(result) => Ok(Some(result)),
            Err(e) => {
                warn!("Polars calculation failed for {}: {}, falling back to manual", symbol, e);
                // Fallback to manual calculation
                Ok(Some(self.calculate_manually(&symbol_key, basis_data, price_data)))
            }
        }
    }

    /// Calculate indicators using Polars (optimized).
    fn calculate_with_polars(
        &self,
        symbol: &str,
        basis_data: &[f64],
        price_data: &[f64],
    ) -> Result<IndicatorResult> {
        // Convert to Polars Series only when needed (performance optimization)
        let basis_series = Series::new("basis", basis_data);
        let price_series = Series::new("price", price_data);

        // Get the last window_size values
        let basis_window = basis_series.tail(Some(self.window_size));
        let price_window = price_series.tail(Some(self.window_size));

        // Calculate MA (simple moving average)
        let ma_basis = basis_window.mean().unwrap_or(0.0);
        let ma_price = price_window.mean().unwrap_or(0.0);

        // Calculate EMA (exponential moving average)
        let ema_basis = self.calculate_ema(basis_data);

        // Calculate Z-Score: (current - MA) / StdDev
        let current_basis = *basis_data.last().unwrap_or(&0.0);
        // Manual std: sqrt(variance) - use basis_data slice directly
        let n = basis_data.len().min(self.window_size);
        let start = basis_data.len().saturating_sub(n);
        let window: Vec<f64> = basis_data[start..].to_vec();
        let mean = ma_basis;
        let variance = if window.len() > 1 {
            window.iter().map(|x| (x - mean).powi(2)).sum::<f64>()
                / (window.len() - 1) as f64
        } else {
            0.0
        };
        let std_basis = variance.sqrt();

        let z_score = if std_basis > 0.0 {
            (current_basis - ma_basis) / std_basis
        } else {
            0.0
        };

        Ok(IndicatorResult {
            symbol: symbol.to_string(),
            timestamp: std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH)
                .unwrap_or_default()
                .as_secs(),
            ma_price,
            ma_basis,
            ema_basis,
            z_score,
        })
    }

    /// Calculate indicators manually (fallback).
    fn calculate_manually(
        &self,
        symbol: &str,
        basis_data: &[f64],
        price_data: &[f64],
    ) -> IndicatorResult {
        // Get the last window_size values
        let basis_window = &basis_data[basis_data.len() - self.window_size..];
        let price_window = &price_data[price_data.len() - self.window_size..];

        // Calculate MA (simple moving average)
        let ma_basis: f64 = basis_window.iter().sum::<f64>() / basis_window.len() as f64;
        let ma_price: f64 = price_window.iter().sum::<f64>() / price_window.len() as f64;

        // Calculate EMA (exponential moving average)
        let ema_basis = self.calculate_ema(basis_window);

        // Calculate Z-Score: (current - MA) / StdDev
        let current_basis = *basis_window.last().unwrap_or(&0.0);

        let variance: f64 = basis_window
            .iter()
            .map(|&x| (x - ma_basis).powi(2))
            .sum::<f64>()
            / basis_window.len() as f64;

        let std_basis = variance.sqrt();

        let z_score = if std_basis > 0.0 {
            (current_basis - ma_basis) / std_basis
        } else {
            0.0
        };

        IndicatorResult {
            symbol: symbol.to_string(),
            timestamp: std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH)
                .unwrap_or_default()
                .as_secs(),
            ma_price,
            ma_basis,
            ema_basis,
            z_score,
        }
    }

    /// Calculate EMA using the current alpha.
    fn calculate_ema(&self, data: &[f64]) -> f64 {
        if data.is_empty() {
            return 0.0;
        }

        // Use the last window_size values
        let window_data = if data.len() >= self.window_size {
            &data[data.len() - self.window_size..]
        } else {
            data
        };

        // Calculate EMA manually
        let mut ema = window_data[0];
        for &value in &window_data[1..] {
            ema = self.alpha * value + (1.0 - self.alpha) * ema;
        }

        ema
    }

    /// Calculate indicators for all symbols with sufficient data.
    pub fn calculate_all_indicators(&self) -> HashMap<String, IndicatorResult> {
        let mut results = HashMap::new();

        for symbol in self.basis_history.keys() {
            if let Ok(Some(result)) = self.calculate_indicators(symbol) {
                results.insert(symbol.clone(), result);
            }
        }

        results
    }

    /// Get the number of symbols with indicator data.
    pub fn symbol_count(&self) -> usize {
        self.basis_history.len()
    }

    /// Get the window size.
    pub fn window_size(&self) -> usize {
        self.window_size
    }
}

impl Default for IndicatorCalculator {
    fn default() -> Self {
        Self::new(30)
    }
}